exporter/awsemfexporter/cwlog_client.go

```diff
@@ -0,0 +1,239 @@
+	defaultRetryCount          = 1
+	ErrCodeThrottlingException = "ThrottlingException"
+)
+
+var (
+	// backoff retry 2 times
+	sleeps = []time.Duration{
+		time.Millisecond * 200, time.Millisecond * 400}
+)
+
+//The log client will perform the necessary operations for publishing log events use case.
+type LogClient interface {
+	PutLogEvents(input *cloudwatchlogs.PutLogEventsInput, retryCnt int) (*string, error)
+	CreateStream(logGroup, streamName *string) (token string, e error)
+}
+
+// Possible exceptions are combination of common errors (https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/CommonErrors.html)
+// and API specific erros (e.g. https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html#API_PutLogEvents_Errors)
+type cloudWatchLogClient struct {
+	svc    cloudwatchlogsiface.CloudWatchLogsAPI
+	logger *zap.Logger
+}
+
+//Create a log client based on the actual cloudwatch logs client.
+func newCloudWatchLogClient(svc cloudwatchlogsiface.CloudWatchLogsAPI, logger *zap.Logger) *cloudWatchLogClient {
+	logClient := &cloudWatchLogClient{svc: svc,
+		logger: logger}
+	return logClient
+}
+
+// NewCloudWatchLogsClient create cloudWatchLogClient
+func NewCloudWatchLogsClient(logger *zap.Logger, awsConfig *aws.Config, sess *session.Session) LogClient {
+	client := cloudwatchlogs.New(sess, awsConfig)
+	client.Handlers.Build.PushBackNamed(handler.RequestStructuredLogHandler)
+	return newCloudWatchLogClient(client, logger)
+}
+
+//Put log events. The method mainly handles different possible error could be returned from server side, and retries them
+//if necessary.
+func (client *cloudWatchLogClient) PutLogEvents(input *cloudwatchlogs.PutLogEventsInput, retryCnt int) (*string, error) {
+	var response *cloudwatchlogs.PutLogEventsOutput
+	var err error
+	var token = input.SequenceToken
+
+	for i := 0; i <= retryCnt; i++ {
+		input.SequenceToken = token
+		response, err = client.svc.PutLogEvents(input)
+		if err != nil {
+			awsErr, ok := err.(awserr.Error)
+			if !ok {
+				client.logger.Error(fmt.Sprintf("E! Cannot cast PutLogEvents error %#v into awserr.Error.", err))
+				return token, err
+			}
+			switch e := awsErr.(type) {
+			case *cloudwatchlogs.InvalidParameterException:
+				client.logger.Error(fmt.Sprintf("E! cloudwatchlogs: %s for log group %s log stream %s, will not retry the request: %s, original error: %#v, %#v",
+					e.Code(),
+					*input.LogGroupName,
+					*input.LogStreamName,
+					e.Message(),
+					e.Error(),
+					e))
+				return token, err
+			case *cloudwatchlogs.InvalidSequenceTokenException: //Resend log events with new sequence token when InvalidSequenceTokenException happens
+				client.logger.Warn(fmt.Sprintf("W! cloudwatchlogs: %s, will search the next token and retry the request: %s, original error: %#v, %#v",
+					e.Code(),
+					e.Message(),
+					e.Error(),
+					e))
+				backoffSleep(i)
```

Comment:
should it be waited if the error type is InvalidSequenceTokenException? it seems we can retry it right a way with new token?

Reply:
Good point.